lib/benchmarks/gallery_automator.dart

```diff
@@ -0,0 +1,199 @@
+  'Progress indicators@material',
+  'Activity indicator@cupertino',
+  'Colors@other',
+];
+
+const Duration _defaultWaitingDuration = Duration(seconds: 3);
+
+enum DemoType {
+  study,
+  animatedWidget,
+  unanimatedWidget,
+}
+
+DemoType typeOfDemo(String demo) {
+  if (demo.contains('@study')) {
+    return DemoType.study;
+  } else if (_demosWithAnimation.contains(demo)) {
+    return DemoType.animatedWidget;
+  } else {
+    return DemoType.unanimatedWidget;
+  }
+}
+
+/// A class that automates the gallery.
+class GalleryAutomator {
+  GalleryAutomator({
+    @required this.benchmarkName,
+    @required this.reporter,
+    @required this.reportError,
+    @required this.runCriterion,
+    this.testScrollsOnly = false,
+  });
+
+  final String benchmarkName;
+  final void Function(String) reporter;
+  final Future<void> Function(dynamic error, StackTrace stackTrace) reportError;
+  final bool Function(String) runCriterion;
+  final bool testScrollsOnly;
+
+  bool finished = false;
+  LiveWidgetController controller;
+
+  /// An iterable that generates all demo names.
+  Iterable<String> get demoNames =>
+      allGalleryDemos(GalleryLocalizationsEn()).map((demo) => demo.describe);
+
+  Future<void> _handleError(dynamic error, StackTrace stackTrace) async {
+    print('Error: $error');
+    print('StackTrace: $stackTrace');
+    await reportError(error, stackTrace);
+  }
+
+  Widget createWidget() {
+    Future<void>.delayed(
+      _initialWaitingDuration,
+      testScrollsOnly ? automateScrolls : automateDemoGestures,
+    ).catchError(_handleError);
+    return const GalleryApp();
+  }
+
+  /// Opens and quits demos that are specified by [runCriterion], twice.
+  Future<void> automateDemoGestures() async {
+    await animationStops();
+
+    controller = LiveWidgetController(WidgetsBinding.instance);
+
+    reporter('==== List of demos to be run ====');
+    for (final demo in demoNames) {
+      if (!runCriterion(demo)) continue;
+      reporter(demo);
+    }
+    reporter('==== End of list of demos to be run ====');
+
+    var finishedStudyDemos = false;
+
+    for (final demo in demoNames) {
+      if (!finishedStudyDemos && typeOfDemo(demo) != DemoType.study) {
+        finishedStudyDemos = true;
+
+        await scrollUntilVisible(
+          element: find.text('Categories').evaluate().single,
+          strict: true,
+          animated: false,
+        );
+      }
+
+      final demoButton =
+          find.byKey(ValueKey(demo), skipOffstage: false).evaluate().single;
+
+      await scrollUntilVisible(
+        element: demoButton,
+        animated: false,
+        strict: true,
+      );
+
+      // Skip demo if it does not pass `runCriterion`.
+      // This continue statement is placed here because we need to scroll
+      // through each `Scrollable` to find the `demoButton`.
+      if (!runCriterion(demo)) continue;
+
+      reporter('Running demo "$demo"');
+
+      for (var i = 0; i < 2; ++i) {
```

Comment:
Where does two come from and if this is always going to be doesn't this only run one time and doesn't necessarily have to be used

Reply:
We open and close every demo twice, in order to measure initial and subsequent builds.
Similar to https://github.com/flutter/gallery/blob/master/test_driver/transitions_perf_test.dart#L157